pkg/apis/raven/v1alpha1/gateway_types.go

```diff
@@ -32,60 +29,94 @@ const (
 	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
-	// TODO add a field to configure using vxlan or host-gw for inner gateway communication?
-	// Endpoints is a list of available Endpoint.
-	Endpoints []Endpoint `json:"endpoints"`
+	// ProxyConfig determine the l7 proxy configuration
+	ProxyConfig ProxyConfiguration `json:"proxyConfig,omitempty"`
+	// TunnelConfig determine the l3 tunnel configuration
+	TunnelConfig TunnelConfiguration `json:"tunnelConfig,omitempty"`
+	// Endpoints are a list of available Endpoint.
+	Endpoints []Endpoint `json:"endpoints,omitempty"`
 	// ExposeType determines how the Gateway is exposed.
-	ExposeType ExposeType `json:"exposeType,omitempty"`
+	ExposeType string `json:"exposeType,omitempty"`
 }
 
-// Endpoint stores all essential data for establishing the VPN tunnel.
-// TODO add priority field?
+// Endpoint stores all essential data for establishing the VPN tunnel and Proxy
 type Endpoint struct {
 	// NodeName is the Node hosting this endpoint.
-	NodeName string            `json:"nodeName"`
-	UnderNAT bool              `json:"underNAT,omitempty"`
-	PublicIP string            `json:"publicIP,omitempty"`
-	Config   map[string]string `json:"config,omitempty"`
+	NodeName string `json:"nodeName"`
+	// Type is the service type of the node, proxy or tunnel
+	Type string `json:"type"`
+	// Port is the exposed port of the node, proxy=10262 tunnel=4500
+	Port int `json:"port,omitempty"`
+	// UnderNAT indicates whether node is under NAT
+	UnderNAT bool `json:"underNAT,omitempty"`
+	// PublicIP is the exposed IP of the node
+	PublicIP string `json:"publicIP,omitempty"`
+	// Config is a map to record config for the raven agent of node
+	Config map[string]string `json:"config,omitempty"`
 }
 
 // NodeInfo stores information of node managed by Gateway.
 type NodeInfo struct {
-	NodeName  string   `json:"nodeName"`
-	PrivateIP string   `json:"privateIP"`
-	Subnets   []string `json:"subnets"`
+	// NodeName is the Node host name.
+	NodeName string `json:"nodeName"`
+	// PrivateIP is the node private ip address
+	PrivateIP string `json:"privateIP"`
+	// Subnets is the pod ip range of the node
+	Subnets []string `json:"subnets"`
 }
 
 // GatewayStatus defines the observed state of Gateway
 type GatewayStatus struct {
 	// Nodes contains all information of nodes managed by Gateway.
 	Nodes []NodeInfo `json:"nodes,omitempty"`
-	// ActiveEndpoint is the reference of the active endpoint.
-	ActiveEndpoint *Endpoint `json:"activeEndpoint,omitempty"`
+	// ActiveEndpoints is the reference of the active endpoint.
+	ActiveEndpoints []*Endpoint `json:"activeEndpoints,omitempty"`
 }
 
 // +genclient
 // +k8s:openapi-gen=true
 // +kubebuilder:object:root=true
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:scope=Cluster,path=gateways,shortName=gw,categories=all
-//+kubebuilder:printcolumn:name="ActiveEndpoint",type=string,JSONPath=`.status.activeEndpoint.nodeName`
```

Comment:
without this line, `ActiveEndponts` will not display when you get it by `kubectl`

Reply:
The gateway supports multiple ActiveEndpoints, therefore, how to display them is not considered for the time being